### ISSUE TITLE
Feature request - allow setting Encoder/Decoder `userInfo` keys

### ISSUE THREAD

**Issue by samritchie** [groue/GRDB.swift#324]
## What did you do?

Used Persistable+Codable support to save existing Codable types to the DB (Codable support already implemented for JSON api parsing purposes). I have slight differences in the serialisation format depending on whether something is being downloaded from the server as opposed to persisted locally on disk. Previously (when I was saving data locally in JSON files) I was using the `userInfo` dictionary on the encoder to determine whether something was being serialised/deserialized from the API vs local

## What did you expect to happen?

There’d be some way to set `userInfo` keys on the internal Encoder/Decoder used by GRDB

## What happened instead?

Internal `RowDecoder`, `PersistableEncoder` etc seem to be private and always have empty `userInfo` dictionaries. Would be great to have some way to set this. I can check for the absence of my options key in `userInfo`, but I’ve coded all my types to trap if the key is not set, on the assumption an unconfigured JSONEncoder/Decoder has been created somewhere.

## Environment

**GRDB flavor(s):** GRDB
**GRDB version:** v2.9.0
**Installation method:** Carthage
**Xcode version:** 9.2
**Swift version:** 4.0
**Platform(s) running GRDB:** iOS
**macOS version running Xcode:** 10.13.3

**Comment by groue** [comment]
Hello @samritchie,

I agree that GRDB currently uses Codable as a mere convenience for record protocols, and that no real attempt has been made at supporting the whole feature set of Swift encoders. As you have noticed, internal encoders and decoders are inaccessible, and impossible to configure.

If a record type provides its own `encode(to container: inout PersistenceContainer)` method, GRDB won't use its conformance to Codable. This may be your best option today: you won't profit from compiler code generation, but you'll be able to do exactly what you need.

Now your feature request is interesting.

Would you be able to elaborate on your idea? Maybe write some sample code that would fit your needs? I'd get a better understanding of the context. And you'd help me enhance my knowledge of Codable as well.

**Comment by groue** [comment]
Sam, I'm closing this issue due to inactivity. Feel free to come back when you have a story to tell!

**Comment by samritchie** [comment]
Sorry, it’s on my list to get back to. I’m not sure what else there is to tell though. 

Essentially if you have model types that already implement Codable for JSON API support, it’s a fairly neat low-impact option to save these to a database. However, if you have minor differences between the way something is persisted locally vs from the server, the normal Codable mechanism for accomplishing this is extract data from the encoder/decoder userInfo dictionary inside your custom encode/init implementations and use this to modify serialisation behaviour. There’s no way to set this currently in GRDB.

All I’m really after is a public `DatabaseQueue.codableUserInfo` property that gets supplied to the internal coders/decoders.

**Comment by groue** [comment]
> Essentially if you have model types that already implement Codable for JSON API support, it’s a fairly neat low-impact option to save these to a database. However, if you have minor differences between the way something is persisted locally vs from the server, the normal Codable mechanism for accomplishing this is extract data from the encoder/decoder userInfo dictionary inside your custom encode/init implementations and use this to modify serialisation behaviour.

You are right. This `userInfo` dictionary is the "official" way to customize encoding and decoding, according to the [Swift Archival & Serialization proposal](https://github.com/apple/swift-evolution/blob/master/proposals/0166-swift-archival-serialization.md#dynamic-context-based-behavior).

> There’s no way to set this currently in GRDB.

True.

> All I’m really after is a public `DatabaseQueue.codableUserInfo` property that gets supplied to the internal coders/decoders.

I don't follow you. I think you assume a better knowledge of Codable than I actually have.

I'd really like that we'd discuss about some real Swift code, so that we would work towards a solution that would better fit GRDB as a whole.

For example, since `DatabaseQueue` is only one of the three [concurrency modes](https://github.com/groue/GRDB.swift/blob/master/README.md#concurrency), we'd also to take care of `DatabasePool` and `DatabaseSnapshot`. But that's my job of taking care of such things - I don't need you to have an exceptional knowledge of GRDB.

What I need from you is your help making your feature request live. So please come back with sample code, or a pull request, so that I better understand 1. your need, in context, and 2. your suggested solution. Your contribution will profit to you and all GRDB users.

Re-open this issue when you're ready! I'm looking forward to it.

**Comment by groue** [comment]
@samritchie I plan to address your request in the next release. For the context: https://twitter.com/khanlou/status/1027608574929903617

**Comment by groue** [comment]
Since the GRDB encoder/decoders are *hidden* inside the internal implementation of FetchableRecord/PersistableRecord, I'm not sure clients will be able to provide their own userInfo. Maybe they will have to check for a GRDB-defined key:

```swift
struct Player: Decodable, FetchableRecord {
    init(from decoder: Decoder) throws {
        if decoder.userInfo[GRDB.codingKey] != nil {
            // GRDB decoder
        } else {
            // other decoder
        }
    }
}

let players = try Player.fetchAll(db)
```

Another possible solution could be to use a new [row adapter](https://github.com/groue/GRDB.swift/blob/master/README.md#row-adapters). This one could let you provide your custom user info dictionary, with as much context as you need. It could also provide a nice answer to @fpillet's [request](https://github.com/groue/GRDB.swift/pull/313#issuecomment-386882401) for context-aware row decoding:

You would write:

```swift
extension CodingUserInfoKey {
    public static let context: CodingUserInfoKey = CodingUserInfoKey(rawValue: "context")!
}

let adapter = CodingRowAdapter(userInfo: [.context: "foo"])
let players = try request.fetchAll(db, adapter: adapter)
```

This userInfo dictionary would be available right in your Decodable initializer:

```swift
struct Player: Decodable, FetchableRecord {
    init(from decoder: Decoder) throws {
        switch decoder.userInfo[.context] as? String {
        case "foo"?:
            // foo decoding
        case "bar"?:
            // bar decoding
        default:
            // other decoder
        }
    }
}
```

This userInfo dictionary would also be available in the regular `init(row:)` initializer from FetchableRecord, because I would never restrict such a powerful feature to Codable types only:

```swift
struct Player: FetchableRecord {
    init(row: Row) throws {
        switch row.userInfo[.context] as? String {
        case "foo"?:
            // foo decoding
        case "bar"?:
            // bar decoding
        default:
            // other decoder
        }
    }
}
```

OMG couldn't this be a solution to a lagging GRDB issue, which is the inability to provide strategies for encoding values such as dates or uuids?

```swift
struct Player: Codable, FetchableRecord, PersistableRecord {
    static let databaseDateEncodingStrategy = .secondsSince1970 // override default .sqliteIso8601 strategy
    static let databaseUUIDEncodingStrategy = .string // override default .data strategy
    
    var uuid: UUID
    var name: String
    var registrationDate: Date
}
```

OK I see now that there may be more work than expected, but that we could address a lot of use cases in one single well-focused strike.

Another big big challenge is providing support for coding keys in *encoding* as well.

This issue will remain open until a proper solution is found. Contributions are welcome!

**Comment by groue** [comment]
A big thank you to @khanlou for triggering inspiration 👍

**Comment by groue** [comment]
Obsoleted by PR #399

**Comment by groue** [comment]
userInfo customization has shipped in v3.3.0-beta1
